Danihodovic outbound webhooks resource controller

grosser opened the pull request:
@zendesk/bre 

replaces https://github.com/zendesk/samson/pull/3178
@danihodovic 

Now showing inline errors on the page too:
<img width="692" alt="screen shot 2019-02-16 at 8 51 42 am" src="https://user-images.githubusercontent.com/11367/52902698-44903600-31c9-11e9-8b35-6e5f4b3cf54c.png">

danihodovic committed:
Outbound webhooks to use ResourceController

grosser committed:
fixes

grosser commented on app/controllers/outbound_webhooks_controller.rb:
```diff
@@ -1,34 +1,28 @@
-      @new_outbound_webhook = webhook
-      render 'webhooks/index'
-    end
+    super(template: 'webhooks/index')
```
FYI new pattern I'm starting since I'm getting sick of all these 1-off helper methods ... this seems to work fine and means we have less methods flying around

grosser commented on app/controllers/outbound_webhooks_controller.rb:
```diff
@@ -1,34 +1,28 @@
 
-  def outbound_webhook_params
-    params.require(:outbound_webhook).permit(:stage_id, :project_id, :url, :username, :password)
+  def resource_params
```
project_id was never sent in and should not be sent in

grosser commented:
... also added json index for both and made sure not to send the passwords out :D

danihodovic commented:
This is not working as expected
```
$ echo '{"source": "mysrc", "branch": "master", "stage_id": 1}'| http --headers POST 'localhost:9080/projects/be5343b7-d558-5385-8be5-f314a8ef972f/webhooks.json' "Authorization: Bearer token"

HTTP/1.1 500 Internal Server Error
```

Samson logs
```
Completed 500 Internal Server Error in 44ms (ActiveRecord: 1.4ms)



NoMethodError (undefined method `admin_for?' for nil:NilClass):

app/models/access_control.rb:14:in `can?'
app/controllers/concerns/current_user.rb:74:in `can?'
app/controllers/concerns/current_user.rb:61:in `authorize_resource!'
```

Suspect change: https://github.com/zendesk/samson/pull/3179/files#diff-f1be353a0a407c7748e57e2e397809fdR7

If you modify that to
```ruby

```

grosser commented:
yeah that's because the filters are now infront of the `login user` filter :(

grosser commented:
FYI tried doing some "insert action at location" magic but that did not work out :(

```
  def self.insert_before_action(filter, **options, before:)
    chain = __callbacks[:process_action].send(:chain)
    location = chain.index { |callback| callback.filter == before } || raise("Did not find #{before} in the chain")
    chain.insert(location, ActiveSupport::Callbacks::Callback.build(filter))
  end
```

grosser committed:
action callback chains are hard to manipulate, so better to repeat a bit

Tried to hack around it with this, but did not work
another approach would be to define both and then just swap them into place after,
but still kinda hacky

```Ruby
  def self.insert_before_action(filter, **options, before:)
    chain = __callbacks[:process_action].send(:chain)
    location = chain.index { |callback| callback.filter == before } || raise("Did not find #{before} in the chain")
    chain.insert(location, ActiveSupport::Callbacks::Callback.build(filter))
  end
```

grosser commented:
better now ✨

sathishavm reviewed (approved):
Nice!!, Good improvement.
